File: sidecar/kumodeck/analysis/key.py
"""调性识别：chroma → Krumhansl-Schmuckler → 24 调 → Camelot / OpenKey。"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .tempo import stft_magnitude

# 调性只关心频率分辨率，时间分辨率无所谓：
# 契约里 tempo 用的 2048（22050 Hz 下 bin 宽 10.8 Hz）在 180 Hz 以下已经宽过一个半音，
# 低音区会整片糊掉，所以这里把窗加到 4096（bin 宽 5.4 Hz），hop 相应放到 1024 省内存。
N_FFT = 4096
HOP = 1024
# C2 ~ C7，低于 C2 基本是底鼓/低频噪声，高于 C7 是镲片噪声，都会污染 chroma
FMIN = 65.4
FMAX = 2093.0
# 谐波抑制：沿时间的中值滤波长度（帧）
HARMONIC_MEDIAN = 17

# Krumhansl-Schmuckler 模板（契约 3.3 给定，索引 0 = 主音）
MAJOR_PROFILE = (6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88)
MINOR_PROFILE = (6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17)

# 音级 → 显示名。大调用契约表里的写法（Db/Ab/Eb/Bb/F#），
# 小调同样照抄契约表（G#/Ab minor 取 Ab，C#/Db minor 取 Db）。
MAJOR_NAMES = ("C", "Db", "D", "Eb", "E", "F", "F#", "G", "Ab", "A", "Bb", "B")
MINOR_NAMES = ("C", "Db", "D", "Eb", "E", "F", "F#", "G", "Ab", "A", "Bb", "B")

# Camelot 表：逐条照抄契约 3.3 的表格，key = (音级, 是否小调)
# 音级：C=0 Db=1 D=2 Eb=3 E=4 F=5 F#=6 G=7 Ab=8 A=9 Bb=10 B=11
CAMELOT: dict[tuple[int, bool], str] = {
    # ---- A 列：小调 ----
    (8, True): "1A",  # Ab minor / G# minor
    (3, True): "2A",  # Eb minor
    (10, True): "3A",  # Bb minor
    (5, True): "4A",  # F minor
    (0, True): "5A",  # C minor
    (7, True): "6A",  # G minor
    (2, True): "7A",  # D minor
    (9, True): "8A",  # A minor
    (4, True): "9A",  # E minor
    (11, True): "10A",  # B minor
    (6, True): "11A",  # F# minor
    (1, True): "12A",  # Db minor / C# minor
    # ---- B 列：大调 ----
    (11, False): "1B",  # B major
    (6, False): "2B",  # F# major
    (1, False): "3B",  # Db major
    (8, False): "4B",  # Ab major
    (3, False): "5B",  # Eb major
    (10, False): "6B",  # Bb major
    (5, False): "7B",  # F major
    (0, False): "8B",  # C major
    (7, False): "9B",  # G major
    (2, False): "10B",  # D major
    (9, False): "11B",  # A major
    (4, False): "12B",  # E major
}


@dataclass
class KeyResult:
    key: str
    key_short: str
    camelot: str
    open_key: str
    confidence: float
    chroma: list[float] = field(default_factory=list)


def camelot_to_open_key(camelot: str) -> str:
    """Camelot → OpenKey。

    推导：OpenKey 的 1d 定义为 C 大调、1m 为 A 小调，而 Camelot 里 C 大调是 8B、
    A 小调是 8A，两套轮盘只差一个固定旋转量 7。所以 open = camelot − 7，
    结果落到 1..12 循环里，即 ((n − 8) mod 12) + 1；字母 A→m（小调）、B→d（大调）。
    校验：1A(Ab minor) → ((1−8) mod 12)+1 = 6 → 6m ✓；6B(Bb major) → 11d ✓。
    """
    if not camelot or len(camelot) < 2:
        return ""
    letter = camelot[-1].upper()
    try:
        number = int(camelot[:-1])
    except ValueError:
        return ""
    if letter not in ("A", "B") or not 1 <= number <= 12:
        return ""
    open_number = ((number - 8) % 12) + 1
    return f"{open_number}{'m' if letter == 'A' else 'd'}"


def _median_filter_time(spec: np.ndarray, size: int = HARMONIC_MEDIAN) -> np.ndarray:
    """沿时间做中值滤波（谐波成分保留、瞬态打击成分被压掉）。

    分块滑窗：整段一次 sliding_window_view 会把内存放大 size 倍，4 分钟曲子直接上 GB。
    """
    if size <= 1 or spec.shape[1] < size:
        return spec
    pad = size // 2
    padded = np.pad(spec, ((0, 0), (pad, pad)), mode="edge")
    out = np.empty_like(spec)
    block = 512
    for start in range(0, spec.shape[1], block):
        stop = min(start + block, spec.shape[1])
        chunk = padded[:, start : stop + 2 * pad]
        windows = np.lib.stride_tricks.sliding_window_view(chunk, size, axis=1)
        out[:, start:stop] = np.median(windows, axis=-1)
    return out


def chroma_weights(sr: int, n_fft: int = N_FFT) -> tuple[np.ndarray, np.ndarray]:
    """(12, n_selected) 的映射矩阵 + 选中的 bin 掩码。

    每个 bin 按 midi = 69 + 12*log2(f/440) 折算到连续音高，用三角权重摊到相邻两个半音，
    比"四舍五入到最近半音"平滑，能吃掉一点音准偏差和频率量化误差。
    """
    freqs = np.fft.rfftfreq(n_fft, 1.0 / sr)
    mask = (freqs >= FMIN) & (freqs <= FMAX)
    selected = freqs[mask]
    weights = np.zeros((12, selected.size), dtype=np.float32)
    if selected.size == 0:
        return weights, mask

    midi = 69.0 + 12.0 * np.log2(selected / 440.0)
    low = np.floor(midi).astype(int)
    frac = midi - low
    cols = np.arange(selected.size)
    weights[low % 12, cols] += (1.0 - frac).astype(np.float32)
    weights[(low + 1) % 12, cols] += frac.astype(np.float32)
    return weights, mask


def compute_chroma(samples: np.ndarray, sr: int) -> np.ndarray:
    """12 维 chroma（已归一到最大值 1）。"""
    y = np.asarray(samples, dtype=np.float32).ravel()
    if y.size < N_FFT:
        return np.zeros(12, dtype=np.float64)
    peak = float(np.max(np.abs(y)))
    if peak > 0:
        y = y / peak

    spec = stft_magnitude(y, n_fft=N_FFT, hop=HOP)
    weights, mask = chroma_weights(sr, N_FFT)
    spec = spec[mask]
    if spec.size == 0:
        return np.zeros(12, dtype=np.float64)

    spec = _median_filter_time(spec, HARMONIC_MEDIAN)
    frames = (weights @ spec).astype(np.float64)  # (12, n_frames)

    # 每帧 L2 归一，抵消音量起伏；再沿时间取中位数（比均值抗鼓点/瞬态干扰）
    norms = np.linalg.norm(frames, axis=0)
    norms[norms <= 0] = 1.0
    frames = frames / norms
    chroma = np.median(frames, axis=1)

    top = float(chroma.max())
    if top > 0:
        chroma = chroma / top
    return chroma


def _pearson(a: np.ndarray, b: np.ndarray) -> float:
    a = a - a.mean()
    b = b - b.mean()
    denom = float(np.linalg.norm(a) * np.linalg.norm(b))
    if denom <= 0:
        return 0.0
    return float(np.dot(a, b) / denom)


def key_from_chroma(chroma: np.ndarray) -> KeyResult:
    """对 24 个候选调（12 音级 × 大小调）求皮尔逊相关，取最大。"""
    chroma = np.asarray(chroma, dtype=np.float64)
    if chroma.size != 12 or float(np.max(np.abs(chroma))) <= 0:
        return KeyResult(key="", key_short="", camelot="", open_key="", confidence=0.0, chroma=[0.0] * 12)

    major = np.asarray(MAJOR_PROFILE, dtype=np.float64)
    minor = np.asarray(MINOR_PROFILE, dtype=np.float64)
    scores: list[tuple[float, int, bool]] = []
    for tonic in range(12):
        # np.roll(profile, tonic) 把模板的主音位置搬到 tonic 音级上
        scores.append((_pearson(chroma, np.roll(major, tonic)), tonic, False))
        scores.append((_pearson(chroma, np.roll(minor, tonic)), tonic, True))

    scores.sort(key=lambda item: item[0], reverse=True)
    best_score, tonic, is_minor = scores[0]
    second = scores[1][0]
    confidence = float(np.clip((best_score - second) / (abs(best_score) + 1e-9), 0.0, 1.0))

    name = (MINOR_NAMES if is_minor else MAJOR_NAMES)[tonic]
    key = f"{name} {'minor' if is_minor else 'major'}"
    key_short = f"{name}m" if is_minor else name
    camelot = CAMELOT[(tonic, is_minor)]
    return KeyResult(
        key=key,
        key_short=key_short,
        camelot=camelot,
        open_key=camelot_to_open_key(camelot),
        confidence=round(confidence, 3),
        chroma=[round(float(v), 4) for v in chroma],
    )


def analyze_key(samples: np.ndarray, sr: int) -> KeyResult:
    """完整调性分析。samples 是单声道 float32。"""
    return key_from_chroma(compute_chroma(samples, sr))

File: crates/kumodeck-analysis/src/key.rs
//! 调性识别：chroma → Krumhansl-Schmuckler → 24 调 → Camelot / OpenKey。

use crate::dsp::{self, median, rfft_freqs};

/// 调性只关心频率分辨率，时间分辨率无所谓：tempo 用的 2048 在 180 Hz 以下
/// 已经宽过一个半音，低音区会整片糊掉，所以这里窗加到 4096（bin 宽 5.4 Hz）。
const N_FFT: usize = 4096;
const HOP: usize = 1024;
/// C2 ~ C7。低于 C2 基本是底鼓/低频噪声，高于 C7 是镲片噪声，都会污染 chroma。
const FMIN: f64 = 65.4;
const FMAX: f64 = 2093.0;
/// 谐波抑制：沿时间的中值滤波长度（帧）
const HARMONIC_MEDIAN: usize = 17;

/// Krumhansl-Schmuckler 模板（索引 0 = 主音）
const MAJOR_PROFILE: [f64; 12] = [
    6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88,
];
const MINOR_PROFILE: [f64; 12] = [
    6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17,
];

/// 音级 → 显示名。大小调用同一张表（G#/Ab minor 取 Ab，C#/Db minor 取 Db）。
const NAMES: [&str; 12] = [
    "C", "Db", "D", "Eb", "E", "F", "F#", "G", "Ab", "A", "Bb", "B",
];

/// Camelot 轮盘。索引 = 音级（C=0 … B=11），`.0` 是小调、`.1` 是大调。
const CAMELOT: [(&str, &str); 12] = [
    ("5A", "8B"),   // C
    ("12A", "3B"),  // Db
    ("7A", "10B"),  // D
    ("2A", "5B"),   // Eb
    ("9A", "12B"),  // E
    ("4A", "7B"),   // F
    ("11A", "2B"),  // F#
    ("6A", "9B"),   // G
    ("1A", "4B"),   // Ab
    ("8A", "11B"),  // A
    ("3A", "6B"),   // Bb
    ("10A", "1B"),  // B
];

#[derive(Debug, Clone, Default, PartialEq)]
pub struct KeyResult {
    pub key: String,
    pub key_short: String,
    pub camelot: String,
    pub open_key: String,
    pub confidence: f64,
    pub chroma: Vec<f64>,
}

/// Camelot → OpenKey。
///
/// 两套轮盘只差一个固定旋转量 7：Camelot 的 8B 是 C 大调、OpenKey 的 1d 也是 C 大调，
/// 所以 `open = ((n - 8) mod 12) + 1`；字母 A→m（小调）、B→d（大调）。
pub fn camelot_to_open_key(camelot: &str) -> String {
    if camelot.len() < 2 {
        return String::new();
    }
    let (number_part, letter) = camelot.split_at(camelot.len() - 1);
    let letter = letter.to_ascii_uppercase();
    let Ok(number) = number_part.parse::<i32>() else {
        return String::new();
    };
    if !(1..=12).contains(&number) || (letter != "A" && letter != "B") {
        return String::new();
    }
    let open_number = (number - 8).rem_euclid(12) + 1;
    format!("{open_number}{}", if letter == "A" { "m" } else { "d" })
}

/// 沿时间做中值滤波：谐波成分保留、瞬态打击成分被压掉。
fn median_filter_time(spec: &mut dsp::Spectrogram, size: usize) {
    if size <= 1 || spec.frames < size {
        return;
    }
    let pad = size / 2;
    let mut window = vec![0.0f64; size];
    let mut row_out = vec![0.0f32; spec.frames];
    for bin in 0..spec.bins {
        for frame in 0..spec.frames {
            for (slot, offset) in window.iter_mut().zip(0..size) {
                // edge 复制补齐
                let index = (frame + offset).saturating_sub(pad).min(spec.frames - 1);
                *slot = spec.at(bin, index) as f64;
            }
            row_out[frame] = median(&mut window) as f32;
        }
        spec.data[bin * spec.frames..(bin + 1) * spec.frames].copy_from_slice(&row_out);
    }
}

/// `(12, n_selected)` 的映射矩阵 + 选中的 bin 下标。
///
/// 每个 bin 按 `midi = 69 + 12*log2(f/440)` 折算到连续音高，用三角权重摊到相邻两个半音，
/// 比"四舍五入到最近半音"平滑，能吃掉一点音准偏差和频率量化误差。
pub fn chroma_weights(sr: f64, n_fft: usize) -> (Vec<[f32; 12]>, Vec<usize>) {
    let freqs = rfft_freqs(n_fft, sr);
    let selected: Vec<usize> = freqs
        .iter()
        .enumerate()
        .filter(|(_, f)| **f >= FMIN && **f <= FMAX)
        .map(|(i, _)| i)
        .collect();

    let weights = selected
        .iter()
        .map(|index| {
            let mut column = [0.0f32; 12];
            let midi = 69.0 + 12.0 * (freqs[*index] / 440.0).log2();
            let low = midi.floor();
            let frac = midi - low;
            let low_class = (low as i64).rem_euclid(12) as usize;
            let high_class = (low as i64 + 1).rem_euclid(12) as usize;
            column[low_class] += (1.0 - frac) as f32;
            column[high_class] += frac as f32;
            column
        })
        .collect();
    (weights, selected)
}

/// 12 维 chroma（已归一到最大值 1）。
pub fn compute_chroma(samples: &[f32], sr: f64) -> Vec<f64> {
    if samples.len() < N_FFT {
        return vec![0.0; 12];
    }
    let peak = samples.iter().fold(0.0f32, |acc, s| acc.max(s.abs()));
    let normalized: Vec<f32> = if peak > 0.0 {
        samples.iter().map(|s| s / peak).collect()
    } else {
        samples.to_vec()
    };

    let full = dsp::stft_magnitude(&normalized, N_FFT, HOP);
    let (weights, selected) = chroma_weights(sr, N_FFT);
    if selected.is_empty() || full.frames == 0 {
        return vec![0.0; 12];
    }
    // 只留选中的 bin，后续中值滤波和矩阵乘都在这个小得多的谱上做
    let mut spec = dsp::Spectrogram {
        bins: selected.len(),
        frames: full.frames,
        data: vec![0.0f32; selected.len() * full.frames],
    };
    for (row, bin) in selected.iter().enumerate() {
        let src = &full.data[bin * full.frames..(bin + 1) * full.frames];
        spec.data[row * full.frames..(row + 1) * full.frames].copy_from_slice(src);
    }
    median_filter_time(&mut spec, HARMONIC_MEDIAN);

    // 每帧 L2 归一（抵消音量起伏），再沿时间取中位数（比均值抗鼓点/瞬态干扰）
    let mut per_class: Vec<Vec<f64>> = vec![Vec::with_capacity(spec.frames); 12];
    for frame in 0..spec.frames {
        let mut acc = [0.0f64; 12];
        for (row, column) in weights.iter().enumerate() {
            let value = spec.at(row, frame) as f64;
            for class in 0..12 {
                acc[class] += column[class] as f64 * value;
            }
        }
        let norm = acc.iter().map(|v| v * v).sum::<f64>().sqrt();
        let norm = if norm > 0.0 { norm } else { 1.0 };
        for class in 0..12 {
            per_class[class].push(acc[class] / norm);
        }
    }

    let mut chroma: Vec<f64> = per_class.iter_mut().map(|values| median(values)).collect();
    let top = chroma.iter().cloned().fold(0.0f64, f64::max);
    if top > 0.0 {
        for value in chroma.iter_mut() {
            *value /= top;
        }
    }
    chroma
}

fn pearson(a: &[f64], b: &[f64]) -> f64 {
    let mean_a = a.iter().sum::<f64>() / a.len() as f64;
    let mean_b = b.iter().sum::<f64>() / b.len() as f64;
    let mut dot = 0.0;
    let mut norm_a = 0.0;
    let mut norm_b = 0.0;
    for (x, y) in a.iter().zip(b) {
        let dx = x - mean_a;
        let dy = y - mean_b;
        dot += dx * dy;
        norm_a += dx * dx;
        norm_b += dy * dy;
    }
    let denom = norm_a.sqrt() * norm_b.sqrt();
    if denom <= 0.0 {
        0.0
    } else {
        dot / denom
    }
}

/// `np.roll(profile, tonic)`：把模板的主音位置搬到 tonic 音级上。
fn roll(profile: &[f64; 12], tonic: usize) -> [f64; 12] {
    let mut out = [0.0f64; 12];
    for i in 0..12 {
        out[(i + tonic) % 12] = profile[i];
    }
    out
}

/// 对 24 个候选调（12 音级 × 大小调）求皮尔逊相关，取最大。
pub fn key_from_chroma(chroma: &[f64]) -> KeyResult {
    if chroma.len() != 12 || chroma.iter().fold(0.0f64, |acc, v| acc.max(v.abs())) <= 0.0 {
        return KeyResult {
            chroma: vec![0.0; 12],
            ..Default::default()
        };
    }
    let mut scores: Vec<(f64, usize, bool)> = Vec::with_capacity(24);
    for tonic in 0..12 {
        scores.push((pearson(chroma, &roll(&MAJOR_PROFILE, tonic)), tonic, false));
        scores.push((pearson(chroma, &roll(&MINOR_PROFILE, tonic)), tonic, true));
    }
    scores.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));

    let (best_score, tonic, is_minor) = scores[0];
    let second = scores[1].0;
    let confidence = ((best_score - second) / (best_score.abs() + 1e-9)).clamp(0.0, 1.0);

    let name = NAMES[tonic];
    let camelot = if is_minor {
        CAMELOT[tonic].0
    } else {
        CAMELOT[tonic].1
    };
    KeyResult {
        key: format!("{name} {}", if is_minor { "minor" } else { "major" }),
        key_short: if is_minor {
            format!("{name}m")
        } else {
            name.to_string()
        },
        camelot: camelot.to_string(),
        open_key: camelot_to_open_key(camelot),
        confidence: (confidence * 1000.0).round() / 1000.0,
        chroma: chroma
            .iter()
            .map(|v| (v * 10_000.0).round() / 10_000.0)
            .collect(),
    }
}

pub fn analyze_key(samples: &[f32], sr: f64) -> KeyResult {
    key_from_chroma(&compute_chroma(samples, sr))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn camelot_table_matches_the_contract() {
        // 逐条对照契约 3.3 的表格
        assert_eq!(CAMELOT[8].0, "1A", "Ab minor");
        assert_eq!(CAMELOT[0].0, "5A", "C minor");
        assert_eq!(CAMELOT[9].0, "8A", "A minor");
        assert_eq!(CAMELOT[0].1, "8B", "C major");
        assert_eq!(CAMELOT[11].1, "1B", "B major");
        assert_eq!(CAMELOT[4].1, "12B", "E major");
    }

    #[test]
    fn every_camelot_slot_is_used_exactly_once() {
        // 打错一个格子就会让和声推荐把两首毫无关系的歌配到一起
        let mut seen: Vec<String> = CAMELOT
            .iter()
            .flat_map(|(minor, major)| [minor.to_string(), major.to_string()])
            .collect();
        seen.sort();
        seen.dedup();
        assert_eq!(seen.len(), 24, "24 个格子必须互不重复");
    }

    #[test]
    fn open_key_conversion_matches_the_worked_examples() {
        // 注释里的两个校验例子
        assert_eq!(camelot_to_open_key("1A"), "6m");
        assert_eq!(camelot_to_open_key("6B"), "11d");
        // C 大调 8B → 1d，A 小调 8A → 1m
        assert_eq!(camelot_to_open_key("8B"), "1d");
        assert_eq!(camelot_to_open_key("8A"), "1m");
    }

    #[test]
    fn open_key_rejects_malformed_input() {
        assert_eq!(camelot_to_open_key(""), "");
        assert_eq!(camelot_to_open_key("13A"), "");
        assert_eq!(camelot_to_open_key("0A"), "");
        assert_eq!(camelot_to_open_key("8C"), "");
        assert_eq!(camelot_to_open_key("xA"), "");
    }

    #[test]
    fn a_pure_profile_recovers_its_own_key() {
        // 把 C 大调模板本身当 chroma 喂进去，必须得到 C major
        let chroma: Vec<f64> = MAJOR_PROFILE.to_vec();
        let got = key_from_chroma(&chroma);
        assert_eq!(got.key, "C major");
        assert_eq!(got.camelot, "8B");
        assert_eq!(got.open_key, "1d");

        // 移到 D 上就应该是 D major
        let shifted: Vec<f64> = roll(&MAJOR_PROFILE, 2).to_vec();
        let got = key_from_chroma(&shifted);
        assert_eq!(got.key, "D major");
        assert_eq!(got.camelot, "10B");
    }

    #[test]
    fn minor_profiles_are_not_confused_with_their_relative_major() {
        // A 小调和 C 大调用的是同一组音，靠模板权重区分，最容易出错的一对
        let chroma: Vec<f64> = roll(&MINOR_PROFILE, 9).to_vec();
        let got = key_from_chroma(&chroma);
        assert_eq!(got.key, "A minor");
        assert_eq!(got.camelot, "8A");
    }

    #[test]
    fn silence_yields_no_key_rather_than_a_wrong_one() {
        let got = key_from_chroma(&vec![0.0; 12]);
        assert_eq!(got.key, "");
        assert_eq!(got.camelot, "");
        assert_eq!(got.confidence, 0.0);
    }

    #[test]
    fn chroma_of_a_pure_tone_lands_on_that_pitch_class() {
        // 440 Hz = A4 → 音级 9
        let sr = 22050.0;
        let samples: Vec<f32> = (0..sr as usize * 2)
            .map(|i| (2.0 * std::f64::consts::PI * 440.0 * i as f64 / sr).sin() as f32)
            .collect();
        let chroma = compute_chroma(&samples, sr);
        let (peak_class, _) = chroma
            .iter()
            .enumerate()
            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap())
            .unwrap();
        assert_eq!(peak_class, 9, "chroma = {chroma:?}");
    }

    #[test]
    fn short_input_returns_zeros_instead_of_panicking() {
        assert_eq!(compute_chroma(&[0.1; 100], 22050.0), vec![0.0; 12]);
    }
}
